applications/sandbox/engine/shaderlib/stdio.shinc

```diff
@@ -0,0 +1,312 @@
+{
+    mat4 viewMatrix;
+    mat4 projectionMatrix;
+    vec3 position;
+    uint index;
+    vec3 viewDirection;
+    #if defined(GEOMETRY_SHADER)
+    vec3 toView[_L_geom_vtxc];
+    #else
+    vec3 toView;
+    #endif
+};
+
+struct CameraInput
+{
+    mat4 view;
+    mat4 proj;
+    vec4 posmeta;
+    vec4 vdirmeta;
+    uint idx;
+};
+
+uniform CameraInput _L_cmr_in : SV_CAMERA;
+
+#define ViewMatrix _L_cmr_in.view;
+#define ProjectionMatrix _L_cmr_in.proj;
+#define ViewProjectionMatrix (_L_cmr_in.proj * _L_cmr_in.view)
+
+vec4 WorldToScreenSpacePosition(vec3 pos)
+{
+    return ((_L_cmr_in.proj * _L_cmr_in.view) * vec4(pos, 1.f));
+}
+
+vec4 WorldToScreenSpacePosition(vec4 pos)
+{
+    return ((_L_cmr_in.proj * _L_cmr_in.view) * pos);
+}
+
+Camera GetCamera()
+{
+    Camera camera;
+    camera.viewMatrix = _L_cmr_in.view;
+    camera.projectionMatrix = _L_cmr_in.proj;
+    camera.position = _L_cmr_in.posmeta.xyz;
+    camera.index = _L_cmr_in.idx;
+    camera.viewDirection = _L_cmr_in.vdirmeta.xyz;
+    #if defined(VERTEX_SHADER)
+    camera.toView = camera.position - (modelMatrix * vec4(position, 1.0)).xyz;
+    #elif defined(FRAGMENT_SHADER)
+    camera.toView = camera.position - _L_stdio.wvertex;
+    #elif defined(GEOMETRY_SHADER)
+    for(int i = 0; i < _L_geom_vtxc; i++)
+        camera.toView[i] = camera.position - _L_stdin[i].wvertex;
+    #endif
+    return camera;
+}
+
+vec4 ModelToScreenSpacePosition(vec3 pos)
+{
+    #if defined(GEOMETRY_SHADER)
+    return (_L_cmr_in.proj * _L_cmr_in.view * stdin(0, modelMatrix)) * vec4(pos, 1.f);
+    #else
+    return (_L_cmr_in.proj * _L_cmr_in.view * stdio(modelMatrix)) * vec4(pos, 1.f);
+    #endif
+}
+
+#if defined(GEOMETRY_SHADER)  // Geometry shader
+vec3 GetWorldPosition(int index)
+{
+    return normalize(stdin(index, wvertex));
+}
+
+vec3 GetWorldNormal(int index)
+{
+    return normalize(stdin(index, wnormal));
+}
+
+vec3 GetWorldTangent(int index)
+{
+    vec3 worldNormal = GetWorldNormal(index);
+    return normalize(stdin(index, wtangent) - dot(stdin(index, wtangent), worldNormal) * worldNormal);
+}
+
+vec3 GetWorldTangent(int index, vec3 worldNormal)
+{
+    return normalize(stdin(index, wtangent) - dot(stdin(index, wtangent), worldNormal) * worldNormal);
+}
+
+vec3 GetWorldBitangent(int index)
+{
+    vec3 worldNormal = GetWorldNormal(index);
+	vec3 worldTangent = GetWorldTangent(index, worldNormal);
+	return normalize(cross(worldNormal, worldTangent));
+}
+
+vec3 GetWorldBitangent(int index, vec3 worldNormal)
+{
+	vec3 worldTangent = GetWorldTangent(index, worldNormal);
+	return normalize(cross(worldNormal, worldTangent));
+}
+
+vec3 GetWorldBitangent(vec3 worldNormal, vec3 worldTangent)
+{
+	return normalize(cross(worldNormal, worldTangent));
+}
+
+vec3 GetWorldDirection(vec3 dir)
+{
+    return normalize((stdin(0, modelMatrix) * vec4(dir, 0.0)).xyz);
+}
+
+vec3 GetWorldPoint(vec3 p)
+{
+    return (stdin(0, modelMatrix) * vec4(p, 1.0)).xyz;
+}
+
+mat3 GetTBN(int index)
```

Comment:
what does TBN mean? more descriptive name pls

Reply:
TangentBitangentNormal matrix, it's a very common term within pbr though

Reply:
google could not tell me but fine if its common